Metrics generator crash in 2.10.0

**Describe the bug**
After upgrading from 2.9 to 2.10 metrics-generator it started crashing:
```
panic: Invalid name validation scheme requested: unset
goroutine 797 [running]:
github.com/prometheus/common/model.ValidationScheme.IsValidLabelName(0xc013c06dcf?, {0xc019c6b5c0?, 0x1d?})
        /home/runner/work/tempo/tempo/vendor/github.com/prometheus/common/model/metric.go:203 +0xb4
github.com/prometheus/prometheus/model/relabel.relabel(0xc000039960, 0xc000505900)
        /home/runner/work/tempo/tempo/vendor/github.com/prometheus/prometheus/model/relabel/relabel.go:340 +0x4c5
github.com/prometheus/prometheus/model/relabel.ProcessBuilder(...)
        /home/runner/work/tempo/tempo/vendor/github.com/prometheus/prometheus/model/relabel/relabel.go:290
github.com/prometheus/prometheus/storage/remote.(*QueueManager).StoreSeries(0xc0020be000, {0xc01c800000, 0x874e, 0x41d634?}, 0x0)
        /home/runner/work/tempo/tempo/vendor/github.com/prometheus/prometheus/storage/remote/queue_manager.go:999 +0x30d
github.com/prometheus/prometheus/tsdb/wlog.(*Watcher).readSegment(0xc0020a2780, 0xc00401e000, 0x0, 0x1)
        /home/runner/work/tempo/tempo/vendor/github.com/prometheus/prometheus/tsdb/wlog/watcher.go:520 +0x548
github.com/prometheus/prometheus/tsdb/wlog.(*Watcher).readAndHandleError(0xc0020a2780, 0xc00401e000, 0x0, 0x1, 0x7fffffffffffffff)
        /home/runner/work/tempo/tempo/vendor/github.com/prometheus/prometheus/tsdb/wlog/watcher.go:351 +0x4a
github.com/prometheus/prometheus/tsdb/wlog.(*Watcher).watch(0xc0020a2780, 0x0, 0x1)
        /home/runner/work/tempo/tempo/vendor/github.com/prometheus/prometheus/tsdb/wlog/watcher.go:443 +0x645
github.com/prometheus/prometheus/tsdb/wlog.(*Watcher).Run(0xc0020a2780)
        /home/runner/work/tempo/tempo/vendor/github.com/prometheus/prometheus/tsdb/wlog/watcher.go:319 +0x55a
github.com/prometheus/prometheus/tsdb/wlog.(*Watcher).loop(0xc0020a2780)
        /home/runner/work/tempo/tempo/vendor/github.com/prometheus/prometheus/tsdb/wlog/watcher.go:268 +0x118
created by github.com/prometheus/prometheus/tsdb/wlog.(*Watcher).Start in goroutine 447
        /home/runner/work/tempo/tempo/vendor/github.com/prometheus/prometheus/tsdb/wlog/watcher.go:242 +0xed

```

**To Reproduce**
Steps to reproduce the behavior:
1. Install Tempo with tempo-distributed helm chart, I think this is the relevant section:
```
        metricsGenerator:
          enabled: true
          replicas: 4
          resources:
            requests:
              memory: 3Gi
              cpu: 1000m
          config:
            processor:
              service_graphs:
                max_items: 100000
                dimensions:
                  - deployment.environment
              span_metrics:
                intrinsic_dimensions:
                  status_code: true
                dimensions:
                  - deployment.environment
                  - http.status_code
                  - http.route
                  - rpc.system
                  - rpc.service
                  - rpc.method
                  - rpc.grpc.status_code
              local_blocks:
                filter_server_spans: false
            storage:
              remote_write:
                - url: "http://mimir-gateway.mimir.svc.cluster.local/api/v1/push"
                  write_relabel_configs:
                    - source_labels: [ deployment_environment ]
                      regex: (.*)
                      target_label: client_deployment_environment
        overrides:
          defaults:
            global:
              max_bytes_per_trace: 25000000
            ingestion:
              burst_size_bytes: 50000000
              rate_limit_bytes: 30000000
              max_traces_per_user: 20000
            metrics_generator:
              processors:
                - service-graphs
                - span-metrics
                - local-blocks
              processor:
                span_metrics:
                  enable_target_info: true
                  target_info_excluded_dimensions:
                    - k8s_node_name
                    - k8s_pod_ip
                    - k8s_pod_name
                    - k8s_pod_start_time
                    - k8s_pod_uid
                    - process_command
                    - process_command_args
                    - process_executable_name
                    - process_executable_path
                    - process_owner
                    - process_pid
```

**Expected behavior**
Don't crash

**Environment:**
 - Infrastructure: Kubernetes (tempo-distributed)
 - Deployment tool: helm

**Additional Context**
No changes besides upgrading

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -15,6 +15,7 @@
 * [BUGFIX] Correct instant query calculation for rate() [#6205](https://github.com/grafana/tempo/pull/6205) (@ruslan-mikhailov)
 * [BUGFIX] Fix live-store deadlock occurring after a complete block failure [#6338](https://github.com/grafana/tempo/pull/6338) (@ruslan-mikhailov)
 * [BUGFIX] generator: fix dimension_mappings and target_info_excluded_dimensions being unconditionally overwritten even when overrides were nil [#6390](https://github.com/grafana/tempo/pull/6390) (@carles-grafana)
+* [BUGFIX] generator: fix panic when `write_relabel_configs` is configured on remote write endpoints [#6396](https://github.com/grafana/tempo/pull/6396) (@carles-grafana)
 
 ### 3.0 Cleanup
 

--- a/modules/generator/config.go
+++ b/modules/generator/config.go
@@ -131,6 +131,10 @@ func (cfg *Config) Validate() error {
 		}
 	}
 
+	if err := cfg.Storage.Validate(); err != nil {
+		return err
+	}
+
 	if !slices.Contains(validCodecs, cfg.Codec) {
 		return fmt.Errorf("invalid codec: %s, valid choices are %s", cfg.Codec, validCodecs)
 	}

--- a/modules/generator/storage/config.go
+++ b/modules/generator/storage/config.go
@@ -2,8 +2,10 @@ package storage
 
 import (
 	"flag"
+	"fmt"
 	"time"
 
+	"github.com/prometheus/common/model"
 	prometheus_config "github.com/prometheus/prometheus/config"
 	"github.com/prometheus/prometheus/tsdb/agent"
 	"github.com/prometheus/prometheus/util/compression"
@@ -24,6 +26,19 @@ type Config struct {
 	// Prometheus remote write config
 	// https://prometheus.io/docs/prometheus/latest/configuration/configuration/#remote_write
 	RemoteWrite []prometheus_config.RemoteWriteConfig `yaml:"remote_write,omitempty"`
+}
+
+// Validate initializes and validates the remote write configurations. This must
+// be called once at startup before any tenant storage instances are created, so
+// that NameValidationScheme is set on all write relabel configs. Without this,
+// relabeling panics with "Invalid name validation scheme requested: unset".
+func (cfg *Config) Validate() error {
+	for i := range cfg.RemoteWrite {
+		if err := cfg.RemoteWrite[i].Validate(model.UTF8Validation); err != nil {
+			return fmt.Errorf("invalid remote write config %q: %w", cfg.RemoteWrite[i].Name, err)
+		}
+	}
+	return nil
 }
 
 func (cfg *Config) RegisterFlagsAndApplyDefaults(string, *flag.FlagSet) {

--- a/modules/generator/storage/config_util_test.go
+++ b/modules/generator/storage/config_util_test.go
@@ -7,8 +7,12 @@ import (
 	"testing"
 
 	prometheus_common_config "github.com/prometheus/common/config"
+	"github.com/prometheus/common/model"
 	prometheus_config "github.com/prometheus/prometheus/config"
+	"github.com/prometheus/prometheus/model/labels"
+	"github.com/prometheus/prometheus/model/relabel"
 	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
 
 	"github.com/grafana/tempo/pkg/util"
 )
@@ -140,6 +144,48 @@ func Test_copyMap(t *testing.T) {
 	assert.Len(t, original, 2)
 	assert.Equal(t, "v2", original["k2"])
 	assert.Equal(t, "", original["k3"])
+}
+
+func Test_generateTenantRemoteWriteConfigs_writeRelabelConfigs(t *testing.T) {
+	logger := slog.New(slog.NewTextHandler(os.Stdout, nil))
+
+	cfg := Config{
+		RemoteWrite: []prometheus_config.RemoteWriteConfig{
+			{
+				URL:     &prometheus_common_config.URL{URL: urlMustParse("http://prometheus-1/api/prom/push")},
+				Headers: map[string]string{},
+				WriteRelabelConfigs: []*relabel.Config{
+					{
+						SourceLabels: model.LabelNames{"deployment_environment"},
+						Separator:    relabel.DefaultRelabelConfig.Separator,
+						Regex:        relabel.DefaultRelabelConfig.Regex,
+						Replacement:  relabel.DefaultRelabelConfig.Replacement,
+						Action:       relabel.Replace,
+						TargetLabel:  "client_deployment_environment",
+					},
+				},
+			},
+		},
+	}
+
+	// Validate once at startup, as the generator does in Config.Validate().
+	require.NoError(t, cfg.Validate())
+
+	result := generateTenantRemoteWriteConfigs(cfg.RemoteWrite, "my-tenant", nil, false, logger, false)
+
+	// Simulate what the Prometheus remote write queue manager does: run the
+	// relabel configs against a label set. Before the fix this panics with
+	// "Invalid name validation scheme requested: unset" because
+	// NameValidationScheme was never initialized on the relabel configs.
+	lbls := labels.FromStrings("deployment_environment", "production")
+
+	var newLbls labels.Labels
+	var keep bool
+	assert.NotPanics(t, func() {
+		newLbls, keep = relabel.Process(lbls, result[0].WriteRelabelConfigs...)
+	})
+	assert.True(t, keep)
+	assert.Equal(t, "production", newLbls.Get("client_deployment_environment"))
 }
 
 func urlMustParse(urlStr string) *url.URL {

--- a/modules/generator/storage/instance.go
+++ b/modules/generator/storage/instance.go
@@ -169,15 +169,16 @@ func (s *storageImpl) watchOverrides() {
 
 			if !headersEqual(s.currentHeaders, newHeaders) || s.sendNativeHistograms != newSendNativeHistograms {
 				s.logger.Info("updating remote write configuration")
-				s.currentHeaders = newHeaders
-				s.sendNativeHistograms = newSendNativeHistograms
 				err := s.remote.ApplyConfig(&prometheus_config.Config{
 					RemoteWriteConfigs: generateTenantRemoteWriteConfigs(s.cfg.RemoteWrite, s.tenantID, newHeaders, s.cfg.RemoteWriteAddOrgIDHeader, s.logger, newSendNativeHistograms),
 				})
 				if err != nil {
 					metricStorageRemoteWriteUpdateFailed.WithLabelValues(s.tenantID).Inc()
-					s.logger.Info("Failed to update remote write configuration. Remote write will continue with configuration", "err", err.Error())
+					s.logger.Warn("Failed to update remote write configuration. Remote write will continue with previous configuration", "err", err.Error())
+					continue
 				}
+				s.currentHeaders = newHeaders
+				s.sendNativeHistograms = newSendNativeHistograms
 			}
 		case <-s.closeCh:
 			return

